src/action-controller/router/builder.cr

```diff
@@ -486,7 +504,9 @@ module ActionController::Route::Builder
                     ){% if !nilable %}.tap { |result|
                       if result.nil?
                         if params.has_key?({{query_param_name}})
```

Comment:
I would use macros here to check if this is a header param to avoid the `elsif` branch
At compile time we can put in the correct if statement based on the parameter type